substrate/frame/staking-async/runtimes/parachain/src/staking.rs

```
@@ -24,47 +24,187 @@ use pallet_election_provider_multi_block::{self as multi_block, SolutionAccuracy
 use pallet_staking_async::UseValidatorsMap;
 use pallet_staking_async_rc_client as rc_client;
 use polkadot_runtime_common::{prod_or_fast, BalanceToU256, U256ToBalance};
+use sp_core::Get;
+use sp_npos_elections::BalancingConfig;
 use sp_runtime::{
 	traits::Convert, transaction_validity::TransactionPriority, FixedPointNumber, FixedU128,
 	SaturatedConversion,
 };
 use xcm::latest::prelude::*;
 
+/// This macro contains all of the variable parameters that we intend to use for Polkadot and
+/// Kusama.
 parameter_types! {
+	/// Number of election pages that we operate upon.
+	///
+	/// * Polkadot: 32 (3.2m snapshot)
+	/// * Kusama: 16 (1.6m snapshot)
+	///
+	/// Reasoning: Both leads to around 700 nominators per-page, yielding the weights in
+	/// https://github.com/paritytech/polkadot-sdk/pull/8704, the maximum of which being around 1mb
+	/// compressed PoV and 2mb uncompressed.
+	///
+	/// NOTE: in principle, there is nothing preventing us from stretching these values further, it
+	/// will only reduce the per-page POVs. Although, some operations like the first snapshot, and
+	/// the last page of export (where we operate on `MaxValidatorSet` validators) will not get any
+	/// better.
+	pub storage Pages: u32 = 4;
+
+	/// * Polkadot: 8 * 32 (245 blocks, 25.6m). Enough time to verify 8 solutions.
+	/// * Kusama: 4 * 16 (64 blocks, 6.4m). Enough time to verify 4 solutions.
```

maybe a nitpick but maybe it's more correct to say: Enough time to verify up to 4 valid solutions. We allow early failure for a solution (e.g. score < min allowed score or if page fails validation) so we can verify in principle more than 4 if some of them are invalid and fail early.

---

Hmm yeah, and I actually now realize that if we fast-fail a solution at the very beginning due to weak score, we fell away from our assumption of "signed phase always has a multiple of `Pages` blocks left". I think the very important clause you added in last week's PR to ` start` signal will handle this well though. Can you check if our test cases capture this?